hashicorp/dev-portal
# feat: rewrite learn urls

## "Ready for Review" checklist




---

## Relevant links



- [Preview link](https://dev-portal-git-ksrewrite-tutorial-urls-hashicorp.vercel.app/waypoint/tutorials) 🔎
- [Asana task](https://app.asana.com/0/1199634971449915/1201956804128036) 🎟️

## What & Why

This PR adds a remark plugin that  rewrites urls for learn tutorial content that reference other learn tutorials or collections.

If the link references a tutorial or collection that is included in the beta-product list, this rewrites the link path to be relative to dev dot.

All other tutorial and collection links are rewritten to point externally to production learn.hashicorp.com

Once dev portal is publicly available, this plugin will be deprecated as all written internal links within the content will be updated in the content itself. This is an interim workaround while we have a divergent state during beta.

### The api route

In order to rewrite tutorial urls properly to the dev portal format, we need extra information not currently included in the URL to get the collection context. In order to access this context, we need to query the API. This PR adds a new api route that creates a tutorial map for all tutorials in the API to ease this look-up. 

## How

See comments for annotation on approach. 

## Testing

- Caching: Visit the [`api/tutorials-map` vercel log page](https://vercel.com/hashicorp/dev-portal/6bscg864YkrEtxwyvwwPMJZaYVKe/functions?name=api%2Ftutorials-map). Then [go to a tutorial](https://dev-portal-i4vqlqr0s-hashicorp.vercel.app/waypoint/tutorials/get-started-kubernetes/get-started-intro) and see the logs. Make sure that the _get all tutorials_ call only happens once. 
- Link rewriting: Visit a tutorial and click on links within the content: 
- Check for those linking to waypoint or vault that stay within dev portal. 
- Check for those linking to other product tutorials or collections linking externally to learn. 
- Check for external links to be untouched. 

## Anything else?

This plugin doesn't account for docs links. We will need to address that in [follow up work. ](https://app.asana.com/0/1201972623923699/1202144257024022)
